tensorflow/core/kernels/stack.cc

```diff
@@ -319,6 +423,40 @@ void StackPopOp::ComputeAsync(OpKernelContext* ctx, DoneCallback done) {
     ctx->set_output(0, value.tensor);
     done();
   }
+
+  // This is the reverse of the heuristic for swapping out, asynchronously
```

Comment:
Can we remove this reverse heuristic for now unless there's a benchmark showing it's beneficial?

Reply:
Actually, this heuristic is essential (as long as swap out keep using the current heuristic).

The reasoning is, in `StackPopOp()` we asynchronously swap in swapped tensors that are only needed in the future, where previously, swapped tensors are only swapped in when needed for computation. This increasing of asynchronicity is the core speedup idea, which is allowed by swapping oldest tensors first.

Thus, the reverse heuristic is needed to keep TensorFlow from swapping in those "future" tensors too aggressively. Since in `StackPushOp()` we swap when the memory is >70% full, we also begin swapping in when the memory is <70% full. (Though, if a needed tensor isn't yet swapped in, it's immediately swapped in, as seen at `StackPopOp()`).

Reply:
Yeah, I see your reasoning. I wonder if we need different kOccupancy thresholds for swapping in/out to avoid too much churn? Can we try to modify the existing benchmarks to measure the difference (it might be difficult to get it to start swapping, though)?

Reply:
In the past I've tried higher kOccupancy for swapping in, but it causes OOMs instead. I haven't tried lower values, though, because since it's a heuristic I'm not sure of the best way to measure it (might see different results for different parameters, etc).

Reply:
By the way, how to benchmark with rnn_long_sequence_benchmark?

I tried this one from https://github.com/tensorflow/tensorflow/blob/master/CONTRIBUTING.md:

```
./tensorflow/tools/ci_build/ci_build.sh gpu bazel test //tensorflow/python/kernel_tests:rnn_test
```

But keep getting "Executed 1 out of 1 test: 1 fails locally", even on `master` branch.

Reply:
See the first answer here:
https://stackoverflow.com/questions/34403873/how-can-i-create-run-benchmarks-for-custom-kernels-in-tensorflow

Make sure to pass --config=cuda so that it runs on GPU.